api/test/context/runtime_context_test.cc

```diff
@@ -0,0 +1,63 @@
+#include "opentelemetry/context/threadlocal_context.h"
+
+#include <gtest/gtest.h>
+
+using namespace opentelemetry;
+
+// Tests that GetCurrent returns the current context
+TEST(RuntimeContextTest, GetCurrent)
+{
+  std::map<std::string, context::ContextValue> map_test = {{"test_key", (int64_t)123}};
+  context::Context test_context                         = context::Context(map_test);
+  context::RuntimeContext::Token old_context = context::RuntimeContext::Attach(&test_context);
```

Comment:
Another thing I wonder about: whenever a new span is started, we have to set the current span. I think we can only do this by adding a new context. So we'll add a new context when a span is started and we end the context when the span is ended. The user though has the option to end a child span before a parent span. This shouldn't screw up our context stack. Maybe I'm also not completely understanding the workflow here, please correct me if I'm wrong.

Reply:
Perhaps RuntimeContext::Attach should take a context object instead of a ptr and that should prevent issues if the original context object goes out of scope.

Reply:
> What is the memory management strategy when managing context? Will contexts always be stored on the stack?

I think the context object has to be always allocated on the heap. Context can be propagated between threads and threads can get killed by the underlying OS (which means the stack will be reclaimed and repurposed).

Reply:
> The user though has the option to end a child span before a parent span. This shouldn't screw up our context stack.

That's a very interesting topic, @pyohannes what would you expect for the following scenario?

```C++
// current span = EMPTY
span("A") begin
// current span = A
span("B") begin
// current span = B
span("C") begin
// current span = C
span("B") end
// current span = ???
span("A") end
// current span = ???
span("C") end
// current span = ???
```

Reply:
This would my preferred way:

```C++
// current span = EMPTY
span("A") begin
// current span = A
span("B") begin
// current span = B
span("C") begin
// current span = C
span("B") end
// current span = C
span("A") end
// current span = C
span("C") end
// current span = EMPTY
```

Reply:
@cijothomas

Reply:
Yes, this is what I would expect. The use case is less exotic than it might sound. For asynchronous operations, one might bind a span to the operation and let the operation end the span when it ends. The parent span can end before the operation is done. This doesn't necessarily involve threads.

Reply:
> I think the context object has to be always allocated on the heap.

Then we need to define who owns the context object and who frees it. It isn't clear to me with the current design.

Reply:
> Then we need to define who owns the context object and who frees it. It isn't clear to me with the current design.

I agree, though I was originally thinking to unblock this PR first and then revisit (as the TLS part is helpful to unblock the HTTP propagation) 😄

Reply:
In the meantime should I stay away from the Attach method accepting a context*?

Reply:
I don't want to block this PR 'coz I wish to unblock the HTTP integration.
From correctness perspective, I think all context objects should be reference so yes eventually we need to stay away from raw ptr.

Reply:
Changed to RuntimeContext accepting a Context object instead of Context* so we should not be ref counting it.

Reply:
*now

Reply:
>  I was originally thinking to unblock this PR first and then revisit

That's ok for me, as long as we keep track of open issues (memory management and asynchronous spans), preferably via GitHub issues.

Reply:
Are we good to merge then?

Reply:
#219 created for tracking.